### PR TITLE
Checkpointing enhancement

### PR DESCRIPTION
**Pull request by Ig-dolci**

### CODE REVIEW

File: pyadjoint/tape.py [hunk 1]
```diff
@@ -745,19 +762,30 @@ def copy(self, blocks=None):
...
 
         with stop_annotating():
-            self._checkpoint = {
-                var: var.saved_output._ad_create_checkpoint()
-                for var in self.checkpointable_state
-            }
+            if checkpointable_state:
+                for var in self.checkpointable_state:
+                    if var.checkpoint is not None:
+                        self._checkpoint[var] = var.checkpoint._ad_create_checkpoint()
```
Comment (Ig-dolci) [inline pyadjoint/tape.py:779]:
```suggestion
                        self._checkpoint[var] = var.checkpoint
```

File: pyadjoint/adjfloat.py [hunk 2]
```diff
@@ -336,7 +336,10 @@ def __init__(self, *args):
             self.add_dependency(dep)
 
     def recompute_component(self, inputs, block_variable, idx, prepared):
-        return self.operator(*(term.saved_output for term in self.terms))
+        output = self.operator(*(term.saved_output for term in self.terms))
+        if isinstance(output, float):
```
Comment (Ig-dolci) [inline pyadjoint/adjfloat.py:340]: Use `self._ad_convert_type`

File: pyadjoint/checkpointing.py [hunk 3]
```diff
@@ -141,28 +141,27 @@ def _(self, cp_action, timestep):
...
+            self.tape.timesteps[timestep - 1].checkpoint(
+                checkpointable_state=_store_checkpointable_state,
+                adj_deps=_store_adj_deps,
+            )
+            # Remove unnecessary variables in working memory from previous steps.
+            for var in self.tape.timesteps[timestep - 1].checkpointable_state:
+                var._checkpoint = var.output._ad_clear_checkpoint(var.checkpoint)
+            for block in self.tape.timesteps[timestep - 1]:
+                for output in block.get_outputs():
+                    output.output._ad_clear_checkpoint(output.checkpoint)
```
Comment (Ig-dolci) [inline pyadjoint/checkpointing.py:164]:
```suggestion
                    output.checkpoint = output.output._ad_clear_checkpoint(output.checkpoint)
```

File: pyadjoint/block.py [hunk 4]
```diff
@@ -95,6 +98,12 @@ def reset_variables(self, types=None):
         for output in self._outputs:
             output.reset_variables(types)
 
+    def reset_adjoint_state(self):
+        """Resets the adjoint state of the block.
+
+        """
```
Comment (Ig-dolci) [inline pyadjoint/block.py:104]:
```suggestion
        """Resets the adjoint state of the block.
        """
```
```suggestion
        """Resets the adjoint state of the block.

        """
```

**Review by dham — APPROVED** [review]

**Pull request merged**